libraries/mbed/targets/hal/TARGET_NORDIC/TARGET_MCU_NRF51822/us_ticker.c

```diff
@@ -14,14 +14,146 @@
+
+#define RTC_UNITS_TO_MICROSECONDS(RTC_UNITS) (((RTC_UNITS) * (uint64_t)1000000) / RTC_CLOCK_FREQ)
+#define MICROSECONDS_TO_RTC_UNITS(MICROS)    ((((uint64_t)(MICROS) * RTC_CLOCK_FREQ) + 999999) / 1000000)
+
+static bool              us_ticker_inited = false;
+static volatile uint32_t overflowCount;                   /**< The number of times the 24-bit RTC counter has overflowed. */
+static volatile bool     us_ticker_callbackPending = false;
+static uint32_t          us_ticker_callbackTimestamp;
+
+static inline void rtc1_enableCompareInterrupt(void)
+{
+    NRF_RTC1->EVTENCLR = RTC_EVTEN_COMPARE0_Msk;
+    NRF_RTC1->INTENSET = RTC_INTENSET_COMPARE0_Msk;
+}
+
+static inline void rtc1_disableCompareInterrupt(void)
+{
+    NRF_RTC1->INTENCLR = RTC_INTENSET_COMPARE0_Msk;
+    NRF_RTC1->EVTENCLR = RTC_EVTEN_COMPARE0_Msk;
+}
+
+static inline void rtc1_enableOverflowInterrupt(void)
+{
+    NRF_RTC1->EVTENCLR = RTC_EVTEN_OVRFLW_Msk;
+    NRF_RTC1->INTENSET = RTC_INTENSET_OVRFLW_Msk;
+}
+
+static inline void rtc1_disableOverflowInterrupt(void)
+{
+    NRF_RTC1->INTENCLR = RTC_INTENSET_OVRFLW_Msk;
+    NRF_RTC1->EVTENCLR = RTC_EVTEN_OVRFLW_Msk;
+}
+
+static inline void invokeCallback(void)
+{
+    us_ticker_callbackPending = false;
+    rtc1_disableCompareInterrupt();
+    us_ticker_irq_handler();
+}
+
+/**
+ * @brief Function for starting the RTC1 timer. The RTC timer is expected to
+ * keep running--some interrupts may be disabled temporarily.
+ */
+static void rtc1_start()
+{
+    NRF_RTC1->PRESCALER = 0; /* for no pre-scaling. */
+
+    rtc1_enableOverflowInterrupt();
+
+    NVIC_SetPriority(RTC1_IRQn, RTC1_IRQ_PRI);
+    NVIC_ClearPendingIRQ(RTC1_IRQn);
+    NVIC_EnableIRQ(RTC1_IRQn);
+
+    NRF_RTC1->TASKS_START = 1;
+    nrf_delay_us(MAX_RTC_TASKS_DELAY);
+}
+
+/**
+ * @brief Function for stopping the RTC1 timer. We don't expect to call this.
+ */
+void rtc1_stop(void)
+{
+    NVIC_DisableIRQ(RTC1_IRQn);
+    rtc1_disableCompareInterrupt();
+    rtc1_disableOverflowInterrupt();
+
+    NRF_RTC1->TASKS_STOP = 1;
+    nrf_delay_us(MAX_RTC_TASKS_DELAY);
+
+    NRF_RTC1->TASKS_CLEAR = 1;
+    nrf_delay_us(MAX_RTC_TASKS_DELAY);
+}
+
+/**
+ * @brief Function for returning the current value of the RTC1 counter.
+ *
+ * @return     Current RTC1 counter as a 32-bit value (even though the underlying counter is 24-bit)
+ */
+static __INLINE uint32_t rtc1_getCounter(void)
+{
+    if (NRF_RTC1->EVENTS_OVRFLW) {
+        overflowCount++;
+        NRF_RTC1->EVENTS_OVRFLW = 0;
+        NRF_RTC1->EVTENCLR      = RTC_EVTEN_OVRFLW_Msk;
+    }
+    return (overflowCount << 24) | NRF_RTC1->COUNTER;
+}
+
+/**
+ * @brief Function for handling the RTC1 interrupt.
+ *
+ * @details Checks for timeouts, and executes timeout handlers for expired timers.
+ */
+void RTC1_IRQHandler(void)
+{
+    if (NRF_RTC1->EVENTS_OVRFLW) {
+        overflowCount++;
+        NRF_RTC1->EVENTS_OVRFLW = 0;
+        NRF_RTC1->EVTENCLR      = RTC_EVTEN_OVRFLW_Msk;
+    }
```

Comment:
I think it does give a (small) chance on errors with really unlucky timing. When no interrupts can be handled (for example because it is in the Ticker interrupt routine), and between the check on the overflow flag and the reading of the timer it overflows it will use the new RTC value with the old overflow value. 

As random example what I did once: https://github.com/mbedmicro/mbed/blob/master/libraries/mbed/targets/hal/TARGET_Freescale/TARGET_K20XX/us_ticker.c#L74. There are probably other options too.